### PR TITLE
[GECO-138] Adding an endpoint in v2 files controller for reprocessing

### PR DESCRIPTION
**Pull request by diya17**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

https://diging.atlassian.net/browse/GECO-138
  
## Anything else the reviewer needs to know?

... describe here; e.g. is there another pull request this pull request depends on? If so, link it here. ...

https://github.com/diging/giles-eco-giles-web/pull/109/

### CODE REVIEW

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2FilesController.java [hunk 1]
```diff
@@ -244,5 +252,15 @@ public ResponseEntity<String> getFile(@PathVariable String fileId,
 
         return new ResponseEntity<String>(HttpStatus.OK);
     }
+    
+    @RequestMapping(value = "/documents/{documentId}/reprocess", method = RequestMethod.POST, produces = "application/json;charset=UTF-8")
+    public ResponseEntity<String> reprocessDocument(@PathVariable("documentId") String documentId) {
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2FilesController.java]: this should go in it's own controller

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/EmailConfig.java [hunk 2]
```diff
@@ -52,7 +52,7 @@ public JavaMailSender javaMailSender() {
 
         Properties javaMailProperties = new Properties();
         javaMailProperties.put("mail.transport.protocol", emailTransportProtocol);
-        javaMailProperties.put("mail.smtp.auth", emailSmtpAuth);
+        javaMailProperties.put("mail.smtp.auth", false);
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/EmailConfig.java]: no, this configuration has to come from the configuration

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentController.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    private ITransactionalDocumentService documentService;
+    
+    @RequestMapping(value = "/api/v2/resources/documents/{documentId}/reprocess", method = RequestMethod.POST, produces = "application/json;charset=UTF-8")
+    public ResponseEntity<String> reprocessDocument(@PathVariable("documentId") String documentId) {
+        IDocument document = documentService.getDocument(documentId);
+        if (document == null) {
+            return new ResponseEntity<String>(HttpStatus.NOT_FOUND);
+        }
+        reprocessingService.reprocessDocument(document);
+        return new ResponseEntity<String>(HttpStatus.OK);
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentController.java]: the response should be similar to uploading a file.

**Review by jdamerow — CHANGES_REQUESTED** [review]
please remove changes to the files that haven't changed.

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/util/impl/ResponseHelper.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import edu.asu.diging.gilesecosystem.septemberutil.properties.MessageType;
+import edu.asu.diging.gilesecosystem.septemberutil.service.ISystemMessageHandler;
+import edu.asu.diging.gilesecosystem.web.api.util.IResponseHelper;
+
+@Component
+public class ResponseHelper implements IResponseHelper {
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/util/impl/ResponseHelper.java:21]: javadoc missing

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/IUserHelper.java [hunk 5]
```diff
@@ -1,13 +1,33 @@
...
+
+        @return {@code true} if the user has permission to access the document, {@code false} otherwise.
+    */
+    public abstract boolean checkUserPermission(IDocument document, CitesphereToken citesphereToken);
+    
+    /**
+        Generates a ResponseEntity with an unauthorized status code (401) and a corresponding error message.
+        @return The generated ResponseEntity containing the unauthorized error message.
+    */
+    public abstract ResponseEntity<String> generateUnauthorizedUserResponse();
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/IUserHelper.java]: this is a class to help with user objects, it should not create response entities.

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/impl/UserHelper.java [hunk 6]
```diff
@@ -49,4 +61,21 @@ public User createUser(Connection<?> connection) {
...
+
+        return username.equals(usernameInSystem);
+    }
+
+    @Override
+    public ResponseEntity<String> generateUnauthorizedUserResponse() {
+        Map<String, String> unauthorizedMsgs = new HashMap<String, String>();
+        unauthorizedMsgs.put("errorMsg", "User is not authorized to check status.");
+        unauthorizedMsgs.put("errorCode", "401");
+        return responseHelper.generateResponse(unauthorizedMsgs, HttpStatus.UNAUTHORIZED);
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/impl/UserHelper.java]: does this need it's own method? why not create the map in the controller like for the other responses?

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentController.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    public ResponseEntity<String> reprocessDocument(@PathVariable("documentId") String documentId, CitesphereToken citesphereToken) {
+        IDocument document = documentService.getDocument(documentId);
+        Map<String, String> msgs = new HashMap<String, String>();
+        if (document == null) {
+            msgs.put("errorMsg", "Document Id: " + documentId + " does not exist.");
+            msgs.put("errorCode", "404");
+            return responseHelper.generateResponse(msgs, HttpStatus.NOT_FOUND);
+        }
+        if (!userHelper.checkUserPermission(document, citesphereToken)) {
+            return userHelper.generateUnauthorizedUserResponse();
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentController.java]: see below. why is this an extra method?

File: giles-eco/src/test/java/edu/asu/diging/gilesecosystem/web/api/util/impl/ResponseHelperTest.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    }
+    
+    @Test
+    public void test_generateResponse_success() {
+        Map<String, String> msgs = new HashMap<String, String>();
+        msgs.put("id", "DOC12345");
+        ResponseEntity<String> response = responseHelper.generateResponse(msgs, HttpStatus.OK);
+        Assert.assertEquals(HttpStatus.OK, response.getStatusCode());
+        Assert.assertTrue(response.getBody().contains("\"id\" : \"DOC12345\""));
+    }
```
Comment (jdamerow) [inline giles-eco/src/test/java/edu/asu/diging/gilesecosystem/web/api/util/impl/ResponseHelperTest.java:50]: needs a test for what happens when you pass an empty map

File: giles-eco/src/test/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    private V2ReprocessDocumentController v2ReprocessDocumentController;
+    
+    private String DOCUMENT_ID = "documentId";
+    private DocumentAccess access = DocumentAccess.PRIVATE;
+    private String FILE_ID = "fileId";
+    private String UPLOAD_ID = "uploadId";
+    private String UPLOAD_PROGRESS_ID = "uploadProgressId";
+    private CitesphereToken citesphereToken = new CitesphereToken("71b9cc36-939d-4e28-89e9-e5bfca1b26c3");
+    
+    IDocument document;
```
Comment (jdamerow) [inline giles-eco/src/test/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentControllerTest.java]: is there a reason this variables is the only not private one?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentController.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    
+    @Value("${giles_check_upload_endpoint_v2}")
+    private String uploadEndpoint;
+    
+    @RequestMapping(value = "/api/v2/resources/documents/{documentId}/reprocess", method = RequestMethod.POST, produces = "application/json;charset=UTF-8")
+    public ResponseEntity<String> reprocessDocument(@PathVariable("documentId") String documentId, CitesphereToken citesphereToken) {
+        IDocument document = documentService.getDocument(documentId);
+        Map<String, String> msgs = new HashMap<String, String>();
+        if (document == null) {
+            msgs.put("errorMsg", "Document Id: " + documentId + " does not exist.");
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentController.java]: should be `"Document with id " + documentId + " does not exist."`

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/impl/UserHelper.java [hunk 11]
```diff
@@ -15,6 +24,9 @@ public class UserHelper implements IUserHelper {
     
     @Autowired
     private IUserManager userManager;
+    
+    @Autowired
+    private IResponseHelper responseHelper;
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/impl/UserHelper.java]: remove

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/impl/UserHelper.java [hunk 12]
```diff
@@ -49,4 +61,13 @@ public User createUser(Connection<?> connection) {
     public String createUsername(String username, String providerId) {
         return username + "_" + providerId;
     }
+    
+    @Override
+    public boolean checkUserPermission(IDocument document, CitesphereToken citesphereToken) {
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/config/impl/UserHelper.java]: since the method return a boolean, the method name should be `is...` or `has...`.

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Should be good after that. Please reopen the PR to develop though, once GECO-63 has been merged.

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/util/IResponseHelper.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+public interface IResponseHelper {
+    /**
+    Generates a ResponseEntity with a JSON body based on the given map of messages and HTTP status.
+
+    @param msgs a map of messages where the keys represent message identifiers and the values represent message content
+
+    @param status the HTTP status code to be set in the response
+
+    @return a ResponseEntity object with JSON content and the specified HTTP status
+    */
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/util/IResponseHelper.java:14]: this is not how javadoc should look like (each line needs to start with `*`.

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentController.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    @RequestMapping(value = "/api/v2/resources/documents/{documentId}/reprocess", method = RequestMethod.POST, produces = "application/json;charset=UTF-8")
+    public ResponseEntity<String> reprocessDocument(@PathVariable("documentId") String documentId, CitesphereToken citesphereToken) {
+        IDocument document = documentService.getDocument(documentId);
+        Map<String, String> msgs = new HashMap<String, String>();
+        if (document == null) {
+            msgs.put("errorMsg", "Document with id: " + documentId + " does not exist.");
+            msgs.put("errorCode", "404");
+            return responseHelper.generateResponse(msgs, HttpStatus.NOT_FOUND);
+        }
+        if (!userHelper.isUserPermittedToAccessDocument(document, citesphereToken)) {
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2ReprocessDocumentController.java]: giles needs to check with citesphere if a user has permission to reprocess a document, like here, not internally:
https://github.com/diging/giles-eco-giles-web/blob/f1d3e3979070cf9816376cd5153c7fe57d9a73cb/giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/v2/V2FilesController.java#L210

(eventually this needs to be part of the authentication process but for now it needs to be done in the controller)

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/core/model/impl/File.java [hunk 15]
```diff
@@ -286,5 +298,15 @@ public IFile clone() {
...
+    
+    @Override
+    public void setOldFileVersionIds(List<String> fileVersionIds) {
+        this.oldFileVersionIds = fileVersionIds;
+    }
+    
+    @Override
+    public List<String> getOldFileVersionIds() {
+        return this.oldFileVersionIds;
+    }
```
Comment (jdamerow) [inline giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/core/model/impl/File.java]: isn't this already done in the other PR I just merged?

**Review by jdamerow — CHANGES_REQUESTED** [review]
I think there are a lot of changes that shouldn't show up anymore since I merged the other PR? If they don't disappear when reopening this PR, you might have to create a new PR.

**Comment by jdamerow** [comment]
also there are conflicts

File: giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/util/IResponseHelper.java (hunk 13 above)
Comment (jdamerow) [reply giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/api/util/IResponseHelper.java:14]: every line should start with `*`, also the empty ones.

File: giles-eco/src/test/java/edu/asu/diging/gilesecosystem/web/api/util/impl/ResponseHelperTest.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
+package edu.asu.diging.gilesecosystem.web.api.util.impl;
```
Comment (jdamerow) [inline giles-eco/src/test/java/edu/asu/diging/gilesecosystem/web/api/util/impl/ResponseHelperTest.java:1]: I would use a library here to assert the json (JSONassert seems to be the typically used one) to avoid failing test cases if one of the future versions changes how the json is formatted and things like that.

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
I haven't tested this, but hopefully it'll work ;)
